sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala

```diff
@@ -65,6 +65,40 @@ case class BoundReference(ordinal: Int, dataType: DataType, nullable: Boolean)
   }
 }
 
+/**
+ * This bound reference points to a parameterized slot in an input tuple. It is used in
+ * common sub-expression elimination. When some common sub-expressions have same structural
+ * but different slots of input tuple, we replace `BoundReference` with this parameterized
+ * version. The slot position is parameterized and is given at runtime.
+ */
+case class ParameterizedBoundReference(parameter: String, dataType: DataType, nullable: Boolean)
```

Comment:
Since this is only used for codegen, how about moving this to `org.apache.spark.sql.catalyst.expressions.codegen`?

Reply:
nit: `parameter` -> variableNameForOrdinal` or `paramNameForOrdinal`?

Reply:
Yeah, ok, I will.